awscli/autoprompt/widgets.py

```diff
@@ -0,0 +1,208 @@
+# language governing permissions and limitations under the License.
+from prompt_toolkit.formatted_text import HTML, to_formatted_text
+from prompt_toolkit.formatted_text.utils import fragment_list_to_text
+from prompt_toolkit.buffer import Buffer
+from prompt_toolkit.document import Document
+from prompt_toolkit.layout import HSplit, Window, VSplit
+from prompt_toolkit.layout.controls import BufferControl
+from prompt_toolkit.layout.dimension import Dimension
+from prompt_toolkit.layout.layout import ConditionalContainer
+from prompt_toolkit.layout.processors import Processor, Transformation
+from prompt_toolkit.widgets import Frame, HorizontalLine
+
+from awscli.autoprompt.filters import (
+    help_section_visible, doc_window_has_focus, search_input_has_focus,
+    input_buffer_has_focus, is_history_mode
+)
+
+
+class FormatTextProcessor(Processor):
+    """This Processor is used to transform formatted text into a useable
+    format inside a ``prompt_toolkit.buffer.Buffer``.
+
+    """
+    def apply_transformation(self, text_input):
+        # https://python-prompt-toolkit.readthedocs.io/en/master/pages/reference.html#module-prompt_toolkit.formatted_text
+        fragments = to_formatted_text(
+            HTML(fragment_list_to_text(text_input.fragments)))
+        return Transformation(fragments)
+
+
+class HelpPanelWidget:
+    DIMENSIONS = {
+        'help_window_width_max': 35,
+        'help_window_width_pref': 25,
+    }
+    style = '<style fg="darkturquoise">'
+    eol = '\n'
+
+    def __init__(self):
+        help_doc_buffer = self.create_help_buffer(
+            'help_doc', self.doc_window_key_binding_text()
+        )
+        self.help_doc_window = self.create_help_window(
+            help_doc_buffer, title='Doc panel help')
+        help_input_buffer = self.create_help_buffer(
+            'help_input', self.input_buffer_key_binding_text()
+        )
+        self.help_input_window = self.create_help_window(
+            help_input_buffer, title='Prompt panel help')
+        self.container = HSplit([
+            ConditionalContainer(
+                self.help_doc_window,
+                help_section_visible & (doc_window_has_focus
+                                        | search_input_has_focus)),
+            ConditionalContainer(
+                self.help_input_window,
+                help_section_visible & input_buffer_has_focus)
+        ])
+
+    def input_buffer_key_binding_text(self):
+        return (
+            f'{self.style}[ENTER]</style> Autocomplete '
+            f'Choice/Execute Command{self.eol}'
+            f'{self.style}[F1]</style> Hide/Show Help{self.eol}'
+            f'{self.style}[F2]</style> Focus on Docs{self.eol}'
+            f'{self.style}[F3]</style> Hide/Show Docs{self.eol}'
+            f'{self.style}[F4]</style> One/Multi column prompt{self.eol}'
+            f'{self.style}[CONTROL+R]</style> On/Off bck-i-search'
+        )
```

Comment:
So instead of having a bunch of these helper methods, what are your thoughts about generalizing these methods into their own classes? So for example, we would have a base class that is a prompt toolkit container and try to parameterize/generalize option in class variables as much as we can:
```python
class BaseHelpView:
    TITLE = None
    STYLE = '<style fg="darkturquoise">'
    KEYS_TO_DESCRIPTION_MAPPING = {}
    CONDITION = None

def __pt_container__(self):
    pass
```
and then we would have specific classes for different views:
```
class InputBufferHelpView:
    ...

class DocWindowHelpView:
   ...
```
I like this approach for the following reasons:
1. It makes it really easy to add new views in the future as we would just need to fill out a new class
2. We make the `HelpPanelWidget` class even more slimmed down and general. For example, with this approach we could then allow `help_views` to be provided to its tantalizer and the `HelpPanelWidget` would not necessarily need to know anything about the different views and have specific methods for them.